File: samples/CameraAccess/CameraAccess/Gemini/GeminiLiveService.swift
import Foundation
import UIKit

enum GeminiConnectionState: Equatable {
  case disconnected
  case connecting
  case settingUp
  case ready
  case error(String)
}

@MainActor
class GeminiLiveService: ObservableObject {
  @Published var connectionState: GeminiConnectionState = .disconnected
  @Published var isModelSpeaking: Bool = false

  var onAudioReceived: ((Data) -> Void)?
  var onTurnComplete: (() -> Void)?
  var onInterrupted: (() -> Void)?
  var onDisconnected: ((String?) -> Void)?

  private var webSocketTask: URLSessionWebSocketTask?
  private var receiveTask: Task<Void, Never>?
  private var connectContinuation: CheckedContinuation<Bool, Never>?
  private let delegate = WebSocketDelegate()
  private var urlSession: URLSession!
  private let sendQueue = DispatchQueue(label: "gemini.send", qos: .userInitiated)

  init() {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    // Create session with delegate to monitor WebSocket lifecycle
    self.urlSession = URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
  }

  func connect() async -> Bool {
    guard let url = GeminiConfig.websocketURL() else {
      connectionState = .error("No API key configured")
      return false
    }

    #if DEBUG
    NSLog("[GeminiLive] Connecting to: \(url.host ?? "unknown")...")
    #endif

    connectionState = .connecting

    // Wait for WebSocket open, then send setup, then wait for setupComplete
    let result = await withCheckedContinuation { (continuation: CheckedContinuation<Bool, Never>) in
      self.connectContinuation = continuation

      // Set up delegate callbacks
      self.delegate.onOpen = { [weak self] protocol_ in
        guard let self else { return }
        #if DEBUG
        NSLog("[GeminiLive] WebSocket opened (protocol: \(protocol_ ?? "none"))")
        #endif
        Task { @MainActor in
          self.connectionState = .settingUp
          self.sendSetupMessage()
          self.startReceiving()
        }
      }

      self.delegate.onClose = { [weak self] code, reason in
        guard let self else { return }
        let reasonStr = reason.flatMap { String(data: $0, encoding: .utf8) } ?? "no reason"
        #if DEBUG
        NSLog("[GeminiLive] WebSocket closed: code=\(code.rawValue), reason=\(reasonStr)")
        #endif
        Task { @MainActor in
          self.resolveConnect(success: false)
          self.connectionState = .disconnected
          self.isModelSpeaking = false
          self.onDisconnected?("Connection closed (code \(code.rawValue): \(reasonStr))")
        }
      }

      self.delegate.onError = { [weak self] error in
        guard let self else { return }
        let msg = error?.localizedDescription ?? "Unknown error"
        #if DEBUG
        NSLog("[GeminiLive] Session error: \(msg)")
        #endif
        Task { @MainActor in
          self.resolveConnect(success: false)
          self.connectionState = .error(msg)
          self.isModelSpeaking = false
          self.onDisconnected?(msg)
        }
      }

      self.webSocketTask = self.urlSession.webSocketTask(with: url)
      self.webSocketTask?.resume()

      // Timeout after 15 seconds
      Task {
        try? await Task.sleep(nanoseconds: 15_000_000_000)
        await MainActor.run {
          self.resolveConnect(success: false)
          if self.connectionState == .connecting || self.connectionState == .settingUp {
            self.connectionState = .error("Connection timed out")
          }
        }
      }
    }

    return result
  }

  func disconnect() {
    receiveTask?.cancel()
    receiveTask = nil
    webSocketTask?.cancel(with: .normalClosure, reason: nil)
    webSocketTask = nil
    delegate.onOpen = nil
    delegate.onClose = nil
    delegate.onError = nil
    connectionState = .disconnected
    isModelSpeaking = false
    resolveConnect(success: false)
  }

  private var audioSendCount = 0

  func sendAudio(data: Data) {
    guard connectionState == .ready else { return }
    audioSendCount += 1
    let count = audioSendCount
    sendQueue.async { [weak self] in
      let base64 = data.base64EncodedString()
      let json: [String: Any] = [
        "realtimeInput": [
          "audio": [
            "mimeType": "audio/pcm;rate=16000",
            "data": base64
          ]
        ]
      ]
      self?.sendJSON(json)
      if count <= 5 || count % 50 == 0 {
        NSLog("[GeminiLive] Sent audio chunk #\(count): \(data.count) bytes")
      }
    }
  }

  func sendVideoFrame(image: UIImage) {
    guard connectionState == .ready else { return }
    sendQueue.async { [weak self] in
      guard let jpegData = image.jpegData(compressionQuality: GeminiConfig.videoJPEGQuality) else {
        NSLog("[GeminiLive] Failed to create JPEG from image")
        return
      }
      let base64 = jpegData.base64EncodedString()
      NSLog("[GeminiLive] Sending video frame: \(jpegData.count) bytes JPEG")
      let json: [String: Any] = [
        "realtimeInput": [
          "video": [
            "mimeType": "image/jpeg",
            "data": base64
          ]
        ]
      ]
      self?.sendJSON(json)
    }
  }

  // MARK: - Private

  private func resolveConnect(success: Bool) {
    if let cont = connectContinuation {
      connectContinuation = nil
      cont.resume(returning: success)
    }
  }

  private func sendSetupMessage() {
    let setup: [String: Any] = [
      "setup": [
        "model": GeminiConfig.model,
        "generationConfig": [
          "responseModalities": ["AUDIO"]
        ],
        "systemInstruction": [
          "parts": [
            ["text": GeminiConfig.systemInstruction]
          ]
        ],
        "realtimeInputConfig": [
          "automaticActivityDetection": [
            "disabled": false
          ]
        ]
      ]
    ]
    sendJSON(setup)
  }

  private func sendJSON(_ json: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: json),
          let string = String(data: data, encoding: .utf8) else {
      return
    }
    webSocketTask?.send(.string(string)) { error in
      if let error {
        #if DEBUG
        NSLog("[GeminiLive] Send error: \(error.localizedDescription)")
        #endif
      }
    }
  }

  private func startReceiving() {
    receiveTask = Task { [weak self] in
      guard let self else { return }
      while !Task.isCancelled {
        guard let task = self.webSocketTask else { break }
        do {
          let message = try await task.receive()
          switch message {
          case .string(let text):
            await self.handleMessage(text)
          case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
              await self.handleMessage(text)
            }
          @unknown default:
            break
          }
        } catch {
          if !Task.isCancelled {
            let reason = error.localizedDescription
            #if DEBUG
            NSLog("[GeminiLive] Receive error: \(reason)")
            #endif
            await MainActor.run {
              self.resolveConnect(success: false)
              self.connectionState = .disconnected
              self.isModelSpeaking = false
              self.onDisconnected?(reason)
            }
          }
          break
        }
      }
    }
  }

  private func handleMessage(_ text: String) async {
    guard let data = text.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return
    }

    #if DEBUG
    let keys = json.keys.joined(separator: ", ")
    NSLog("[GeminiLive] Message keys: \(keys)")
    #endif

    // Setup complete
    if json["setupComplete"] != nil {
      #if DEBUG
      NSLog("[GeminiLive] Setup complete - ready for audio/video")
      #endif
      connectionState = .ready
      resolveConnect(success: true)
      return
    }

    // GoAway - server will close soon
    if let goAway = json["goAway"] as? [String: Any] {
      let timeLeft = goAway["timeLeft"] as? [String: Any]
      let seconds = timeLeft?["seconds"] as? Int ?? 0
      #if DEBUG
      NSLog("[GeminiLive] GoAway, time left: \(seconds)s")
      #endif
      connectionState = .disconnected
      isModelSpeaking = false
      onDisconnected?("Server closing (time left: \(seconds)s)")
      return
    }

    // Server content
    if let serverContent = json["serverContent"] as? [String: Any] {
      #if DEBUG
      let scKeys = serverContent.keys.joined(separator: ", ")
      NSLog("[GeminiLive] serverContent keys: \(scKeys)")
      #endif

      if let interrupted = serverContent["interrupted"] as? Bool, interrupted {
        #if DEBUG
        NSLog("[GeminiLive] Model interrupted")
        #endif
        isModelSpeaking = false
        onInterrupted?()
        return
      }

      if let modelTurn = serverContent["modelTurn"] as? [String: Any],
         let parts = modelTurn["parts"] as? [[String: Any]] {
        #if DEBUG
        NSLog("[GeminiLive] modelTurn with \(parts.count) parts")
        #endif
        for part in parts {
          if let inlineData = part["inlineData"] as? [String: Any],
             let mimeType = inlineData["mimeType"] as? String {
            #if DEBUG
            NSLog("[GeminiLive] Part mimeType: \(mimeType)")
            #endif
            if mimeType.hasPrefix("audio/pcm"),
               let base64Data = inlineData["data"] as? String,
               let audioData = Data(base64Encoded: base64Data) {
              #if DEBUG
              NSLog("[GeminiLive] Audio chunk received: \(audioData.count) bytes")
              #endif
              if !isModelSpeaking {
                isModelSpeaking = true
              }
              onAudioReceived?(audioData)
            }
          } else if let text = part["text"] as? String {
            #if DEBUG
            NSLog("[GeminiLive] Text part: \(text.prefix(100))")
            #endif
          } else {
            #if DEBUG
            let partKeys = part.keys.joined(separator: ", ")
            NSLog("[GeminiLive] Unknown part keys: \(partKeys)")
            #endif
          }
        }
      }

      if let turnComplete = serverContent["turnComplete"] as? Bool, turnComplete {
        #if DEBUG
        NSLog("[GeminiLive] Turn complete")
        #endif
        isModelSpeaking = false
        onTurnComplete?()
      }

      // Log input/output transcription if present
      #if DEBUG
      if let inputTranscription = serverContent["inputTranscription"] as? [String: Any],
         let text = inputTranscription["text"] as? String {
        NSLog("[GeminiLive] Input transcription: \(text)")
      }
      if let outputTranscription = serverContent["outputTranscription"] as? [String: Any],
         let text = outputTranscription["text"] as? String {
        NSLog("[GeminiLive] Output transcription: \(text)")
      }
      #endif
    }
  }
}

// MARK: - WebSocket Delegate

private class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
  var onOpen: ((String?) -> Void)?
  var onClose: ((URLSessionWebSocketTask.CloseCode, Data?) -> Void)?
  var onError: ((Error?) -> Void)?

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    onOpen?(`protocol`)
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    onClose?(closeCode, reason)
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    if let error {
      onError?(error)
    }
  }
}

File: VisionClaw4Droid/app/src/main/java/com/visionclaw/android/service/gemini/GeminiLiveService.kt
package com.visionclaw.android.service.gemini

import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.util.concurrent.Executors

/**
 * WebSocket client for the Gemini Live API.
 * Equivalent to iOS GeminiLiveService.swift.
 */
class GeminiLiveService {

    enum class ConnectionState {
        DISCONNECTED, CONNECTING, SETTING_UP, READY, ERROR
    }

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isModelSpeaking = MutableStateFlow(false)
    val isModelSpeaking: StateFlow<Boolean> = _isModelSpeaking.asStateFlow()

    // Callbacks
    var onAudioReceived: ((ByteArray) -> Unit)? = null
    var onTurnComplete: (() -> Unit)? = null
    var onInterrupted: (() -> Unit)? = null
    var onDisconnected: ((String?) -> Unit)? = null
    var onInputTranscription: ((String) -> Unit)? = null
    var onOutputTranscription: ((String) -> Unit)? = null
    var onToolCall: ((GeminiToolCall) -> Unit)? = null
    var onToolCallCancellation: ((GeminiToolCallCancellation) -> Unit)? = null

    // Latency tracking
    private var lastUserSpeechEnd: Long = 0L
    private var responseLatencyLogged = false

    private var webSocket: WebSocket? = null
    private val client = OkHttpClient.Builder()
        .readTimeout(java.time.Duration.ofSeconds(30))
        .build()
    private val sendExecutor = Executors.newSingleThreadExecutor()
    private var connectContinuation: CancellableContinuation<Boolean>? = null
    private var timeoutJob: Job? = null

    suspend fun connect(): Boolean {
        val url = GeminiConfig.websocketUrl() ?: run {
            _connectionState.value = ConnectionState.ERROR
            _errorMessage.value = "No API key configured"
            return false
        }

        _connectionState.value = ConnectionState.CONNECTING

        return suspendCancellableCoroutine { cont ->
            connectContinuation = cont

            val request = Request.Builder().url(url).build()
            webSocket = client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    _connectionState.value = ConnectionState.SETTING_UP
                    sendSetupMessage()
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    handleMessage(text)
                }

                override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                    resolveConnect(false)
                    _connectionState.value = ConnectionState.DISCONNECTED
                    _isModelSpeaking.value = false
                    onDisconnected?.invoke("Connection closed (code $code: $reason)")
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    val msg = t.localizedMessage ?: "Unknown error"
                    resolveConnect(false)
                    _connectionState.value = ConnectionState.ERROR
                    _errorMessage.value = msg
                    _isModelSpeaking.value = false
                    onDisconnected?.invoke(msg)
                }
            })

            // Timeout after 15 seconds
            timeoutJob = CoroutineScope(Dispatchers.Default).launch {
                delay(15_000)
                resolveConnect(false)
                if (_connectionState.value == ConnectionState.CONNECTING ||
                    _connectionState.value == ConnectionState.SETTING_UP
                ) {
                    _connectionState.value = ConnectionState.ERROR
                    _errorMessage.value = "Connection timed out"
                }
            }
        }
    }

    fun disconnect() {
        timeoutJob?.cancel()
        timeoutJob = null
        webSocket?.close(1000, null)
        webSocket = null
        onToolCall = null
        onToolCallCancellation = null
        _connectionState.value = ConnectionState.DISCONNECTED
        _isModelSpeaking.value = false
        resolveConnect(false)
    }

    fun sendAudio(data: ByteArray) {
        if (_connectionState.value != ConnectionState.READY) return
        sendExecutor.execute {
            val base64 = Base64.encodeToString(data, Base64.NO_WRAP)
            val json = JSONObject().apply {
                put("realtimeInput", JSONObject().apply {
                    put("audio", JSONObject().apply {
                        put("mimeType", "audio/pcm;rate=${GeminiConfig.INPUT_AUDIO_SAMPLE_RATE}")
                        put("data", base64)
                    })
                })
            }
            webSocket?.send(json.toString())
        }
    }

    fun sendVideoFrame(bitmap: Bitmap) {
        if (_connectionState.value != ConnectionState.READY) return
        sendExecutor.execute {
            val baos = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, GeminiConfig.VIDEO_JPEG_QUALITY, baos)
            val base64 = Base64.encodeToString(baos.toByteArray(), Base64.NO_WRAP)
            val json = JSONObject().apply {
                put("realtimeInput", JSONObject().apply {
                    put("video", JSONObject().apply {
                        put("mimeType", "image/jpeg")
                        put("data", base64)
                    })
                })
            }
            webSocket?.send(json.toString())
        }
    }

    fun sendToolResponse(response: JSONObject) {
        sendExecutor.execute {
            webSocket?.send(response.toString())
        }
    }

    // -- Private --

    private fun resolveConnect(success: Boolean) {
        timeoutJob?.cancel()
        val cont = connectContinuation
        connectContinuation = null
        if (cont?.isActive == true) {
            cont.resume(success) {}
        }
    }

    private fun sendSetupMessage() {
        val setup = JSONObject().apply {
            put("setup", JSONObject().apply {
                put("model", GeminiConfig.model)
                put("generationConfig", JSONObject().apply {
                    put("responseModalities", JSONArray().put("AUDIO"))
                    put("thinkingConfig", JSONObject().put("thinkingBudget", 0))
                })
                put("systemInstruction", JSONObject().apply {
                    put("parts", JSONArray().put(JSONObject().put("text", GeminiConfig.systemInstruction)))
                })
                put("tools", JSONArray().put(
                    JSONObject().put("functionDeclarations", ToolDeclarations.allDeclarations())
                ))
                put("realtimeInputConfig", JSONObject().apply {
                    put("automaticActivityDetection", JSONObject().apply {
                        put("disabled", false)
                        put("startOfSpeechSensitivity", "START_SENSITIVITY_HIGH")
                        put("endOfSpeechSensitivity", "END_SENSITIVITY_LOW")
                        put("silenceDurationMs", 500)
                        put("prefixPaddingMs", 40)
                    })
                    put("activityHandling", "START_OF_ACTIVITY_INTERRUPTS")
                    put("turnCoverage", "TURN_INCLUDES_ALL_INPUT")
                })
                put("inputAudioTranscription", JSONObject())
                put("outputAudioTranscription", JSONObject())
            })
        }
        webSocket?.send(setup.toString())
    }

    private fun handleMessage(text: String) {
        val json = try { JSONObject(text) } catch (_: Exception) { return }

        // Setup complete
        if (json.has("setupComplete")) {
            _connectionState.value = ConnectionState.READY
            resolveConnect(true)
            return
        }

        // GoAway
        if (json.has("goAway")) {
            val seconds = json.optJSONObject("goAway")
                ?.optJSONObject("timeLeft")?.optInt("seconds", 0) ?: 0
            _connectionState.value = ConnectionState.DISCONNECTED
            _isModelSpeaking.value = false
            onDisconnected?.invoke("Server closing (time left: ${seconds}s)")
            return
        }

        // Tool call
        GeminiToolCall.fromJson(json)?.let { toolCall ->
            Log.i(TAG, "Tool call received: ${toolCall.functionCalls.size} function(s)")
            onToolCall?.invoke(toolCall)
            return
        }

        // Tool call cancellation
        GeminiToolCallCancellation.fromJson(json)?.let { cancellation ->
            Log.i(TAG, "Tool call cancellation: ${cancellation.ids.joinToString()}")
            onToolCallCancellation?.invoke(cancellation)
            return
        }

        // Server content
        val serverContent = json.optJSONObject("serverContent") ?: return

        if (serverContent.optBoolean("interrupted", false)) {
            _isModelSpeaking.value = false
            onInterrupted?.invoke()
            return
        }

        val modelTurn = serverContent.optJSONObject("modelTurn")
        val parts = modelTurn?.optJSONArray("parts")
        if (parts != null) {
            for (i in 0 until parts.length()) {
                val part = parts.getJSONObject(i)
                val inlineData = part.optJSONObject("inlineData")
                if (inlineData != null) {
                    val mimeType = inlineData.optString("mimeType", "")
                    if (mimeType.startsWith("audio/pcm")) {
                        val base64Data = inlineData.optString("data", "")
                        if (base64Data.isNotEmpty()) {
                            val audioData = Base64.decode(base64Data, Base64.NO_WRAP)
                            if (!_isModelSpeaking.value) {
                                _isModelSpeaking.value = true
                                if (lastUserSpeechEnd > 0 && !responseLatencyLogged) {
                                    val latency = System.currentTimeMillis() - lastUserSpeechEnd
                                    Log.i(TAG, "Latency: ${latency}ms (user speech end -> first audio)")
                                    responseLatencyLogged = true
                                }
                            }
                            onAudioReceived?.invoke(audioData)
                        }
                    }
                }
                val textContent = part.optString("text", "")
                if (textContent.isNotEmpty()) {
                    Log.i(TAG, textContent)
                }
            }
        }

        if (serverContent.optBoolean("turnComplete", false)) {
            _isModelSpeaking.value = false
            responseLatencyLogged = false
            onTurnComplete?.invoke()
        }

        serverContent.optJSONObject("inputTranscription")?.let { tx ->
            val txt = tx.optString("text", "")
            if (txt.isNotEmpty()) {
                Log.i(TAG, "You: $txt")
                lastUserSpeechEnd = System.currentTimeMillis()
                responseLatencyLogged = false
                onInputTranscription?.invoke(txt)
            }
        }
        serverContent.optJSONObject("outputTranscription")?.let { tx ->
            val txt = tx.optString("text", "")
            if (txt.isNotEmpty()) {
                Log.i(TAG, "AI: $txt")
                onOutputTranscription?.invoke(txt)
            }
        }
    }

    companion object {
        private const val TAG = "GeminiLive"
    }
}
